File: android/app/src/main/kotlin/dev/local/androidtvremote/RemoteContract.kt
package dev.local.androidtvremote

import kotlinx.coroutines.flow.StateFlow

enum class RemoteCommand(val keyCode: Int) {
    UP(19),
    DOWN(20),
    LEFT(21),
    RIGHT(22),
    SELECT(23),
    BACK(4),
    HOME(3),
    MENU(82),
    POWER(26),
    VOLUME_UP(24),
    VOLUME_DOWN(25),
    MUTE(164);

    fun supports(action: RemoteKeyAction): Boolean =
        action == RemoteKeyAction.SHORT || this == SELECT
}

enum class RemoteKeyAction {
    SHORT,
    START_LONG,
    END_LONG,
}

enum class VoiceState {
    UNAVAILABLE,
    IDLE,
    STARTING,
    LISTENING,
}

enum class RemotePressBehavior {
    SINGLE,
    LONG_PRESS,
    REPEAT,
}

object RemotePressPolicy {
    const val LONG_PRESS_MILLIS = 400L
    const val REPEAT_DELAY_MILLIS = 400L
    const val REPEAT_INTERVAL_MILLIS = 100L

    fun behaviorFor(command: RemoteCommand): RemotePressBehavior = when (command) {
        RemoteCommand.SELECT -> RemotePressBehavior.LONG_PRESS
        RemoteCommand.UP,
        RemoteCommand.DOWN,
        RemoteCommand.LEFT,
        RemoteCommand.RIGHT,
        RemoteCommand.VOLUME_UP,
        RemoteCommand.VOLUME_DOWN,
        -> RemotePressBehavior.REPEAT

        else -> RemotePressBehavior.SINGLE
    }
}

enum class TvSource { DISCOVERY, MANUAL }

data class TvCandidate(
    val locatorKey: String,
    val name: String,
    val host: String,
    val source: TvSource,
)

data class TvDevice(
    val id: String,
    val name: String,
    val source: TvSource,
)

data class LastTvRecord(
    val device: TvDevice,
    val lastHost: String,
    val bonjourLocatorKey: String?,
    val lastConnectedAt: Long,
    val clientIdentityFingerprint: String,
    val pairingPeerFingerprint: String,
    val remotePeerFingerprint: String,
)

enum class RemoteError {
    NETWORK_UNREACHABLE,
    TV_NOT_FOUND,
    PAIRING_REQUIRED,
    PAIRING_CODE_INVALID,
    PAIRING_REJECTED,
    PAIRING_TIMEOUT,
    TRUST_CHANGED,
    CONNECTION_LOST,
    VOICE_PERMISSION_DENIED,
    VOICE_SESSION_FAILED,
    TEXT_INPUT_FAILED,
    UNKNOWN,
}

sealed interface RemoteState {
    data object Idle : RemoteState
    data class Discovering(val devices: List<TvCandidate>) : RemoteState
    data class Connecting(val candidate: TvCandidate) : RemoteState
    data class NeedsPairing(val candidate: TvCandidate) : RemoteState
    data class Pairing(val candidate: TvCandidate) : RemoteState
    data class Connected(val device: TvDevice) : RemoteState
    data class Reconnecting(val device: TvDevice, val attempt: Int) : RemoteState
    data class Disconnected(val device: TvDevice?) : RemoteState
    data class Failed(
        val device: TvDevice?,
        val reason: RemoteError,
        val recoverable: Boolean,
    ) : RemoteState
}

object RetryPolicy {
    private val delays = longArrayOf(1_000L, 2_000L, 4_000L)

    fun delayMillis(attempt: Int): Long? = delays.getOrNull(attempt - 1)
}

object TrustTupleValidator {
    fun isComplete(record: LastTvRecord, actualClientFingerprint: String): Boolean =
        record.device.id.isNotBlank() &&
            record.device.name.isNotBlank() &&
            record.lastHost.isNotBlank() &&
            record.clientIdentityFingerprint.isNotBlank() &&
            record.pairingPeerFingerprint.isNotBlank() &&
            record.remotePeerFingerprint.isNotBlank() &&
            record.device.id == record.remotePeerFingerprint &&
            record.clientIdentityFingerprint == actualClientFingerprint
}

enum class ForegroundAction {
    NONE,
    CONNECT_REMEMBERED,
    START_DISCOVERY,
}

object ForegroundPolicy {
    fun action(alreadyForeground: Boolean, hasValidPairing: Boolean): ForegroundAction = when {
        alreadyForeground -> ForegroundAction.NONE
        hasValidPairing -> ForegroundAction.CONNECT_REMEMBERED
        else -> ForegroundAction.START_DISCOVERY
    }

    fun allowsAutomaticDiscovery(isForeground: Boolean, hasValidPairing: Boolean): Boolean =
        isForeground && !hasValidPairing
}

interface RemoteController {
    val state: StateFlow<RemoteState>
    val discoveredCandidates: StateFlow<List<TvCandidate>>
    val voiceState: StateFlow<VoiceState>

    suspend fun initialize()
    suspend fun enterForeground()
    suspend fun connect(candidate: TvCandidate)
    suspend fun connectRemembered()
    suspend fun submitPairingCode(code: String)
    suspend fun send(command: RemoteCommand, action: RemoteKeyAction = RemoteKeyAction.SHORT)
    suspend fun startVoice()
    suspend fun stopVoice()
    suspend fun disconnect()
    suspend fun forget()
    suspend fun enterBackground()
    fun close()
}

File: ios/AndroidTVRemote/RemoteContract.swift
import Foundation

enum AppRoute: Equatable, Sendable {
    case fullRemote
    case compactRemote

    static let compactURL = URL(string: "androidtvremote://compact")!

    init?(url: URL) {
        guard url.scheme?.caseInsensitiveCompare("androidtvremote") == .orderedSame,
              url.host?.caseInsensitiveCompare("compact") == .orderedSame else {
            return nil
        }
        self = .compactRemote
    }
}

enum RemoteCommand: String, CaseIterable, Codable, Sendable {
    case up
    case down
    case left
    case right
    case select
    case back
    case home
    case menu
    case power
    case volumeUp
    case volumeDown
    case mute

    var androidKeyCode: Int {
        switch self {
        case .up: 19
        case .down: 20
        case .left: 21
        case .right: 22
        case .select: 23
        case .back: 4
        case .home: 3
        case .menu: 82
        case .power: 26
        case .volumeUp: 24
        case .volumeDown: 25
        case .mute: 164
        }
    }

    func supports(_ action: RemoteKeyAction) -> Bool {
        action == .short || self == .select
    }
}

enum RemoteKeyAction: Equatable, Sendable {
    case short
    case startLong
    case endLong
}

enum RemotePressBehavior: Equatable, Sendable {
    case single
    case longPress
    case repeatWhileHeld
}

enum RemotePressPolicy {
    static let longPressThreshold: TimeInterval = 0.4
    static let repeatDelay: TimeInterval = 0.4
    static let repeatInterval: TimeInterval = 0.1

    static func behavior(for command: RemoteCommand) -> RemotePressBehavior {
        switch command {
        case .select:
            return .longPress
        case .up, .down, .left, .right, .volumeUp, .volumeDown:
            return .repeatWhileHeld
        case .back, .home, .menu, .power, .mute:
            return .single
        }
    }
}

enum TvSource: String, Codable, Sendable {
    case discovery
    case manual
}

struct BonjourLocator: Codable, Equatable, Sendable {
    let domain: String
    let type: String
    let name: String
}

struct TvCandidate: Codable, Equatable, Identifiable, Sendable {
    var id: String { locatorKey }

    let locatorKey: String
    let name: String
    let host: String
    let source: TvSource
}

struct RemoteDevice: Codable, Equatable, Identifiable, Sendable {
    let id: String
    let name: String
    let host: String
    let locator: BonjourLocator?
    let source: TvSource

    init(
        id: String,
        name: String,
        host: String,
        locator: BonjourLocator?,
        source: TvSource = .manual
    ) {
        self.id = id
        self.name = name
        self.host = host
        self.locator = locator
        self.source = source
    }
}

struct LastTvRecord: Codable, Equatable, Sendable {
    let persistentDeviceID: String
    let name: String
    let clientIdentityFingerprint: String
    let pairingPeerFingerprint: String
    let remotePeerFingerprint: String
    let lastHost: String
    let bonjourLocator: BonjourLocator?
    let source: TvSource
    let lastConnectedAt: Date

    init(
        persistentDeviceID: String,
        name: String,
        clientIdentityFingerprint: String,
        pairingPeerFingerprint: String,
        remotePeerFingerprint: String,
        lastHost: String,
        bonjourLocator: BonjourLocator?,
        source: TvSource = .manual,
        lastConnectedAt: Date = .distantPast
    ) {
        self.persistentDeviceID = persistentDeviceID
        self.name = name
        self.clientIdentityFingerprint = clientIdentityFingerprint
        self.pairingPeerFingerprint = pairingPeerFingerprint
        self.remotePeerFingerprint = remotePeerFingerprint
        self.lastHost = lastHost
        self.bonjourLocator = bonjourLocator
        self.source = source
        self.lastConnectedAt = lastConnectedAt
    }

    var isComplete: Bool {
        [persistentDeviceID, name, clientIdentityFingerprint, pairingPeerFingerprint,
         remotePeerFingerprint, lastHost].allSatisfy {
            !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        } && persistentDeviceID == remotePeerFingerprint
    }

    var device: RemoteDevice {
        RemoteDevice(
            id: persistentDeviceID,
            name: name,
            host: lastHost,
            locator: bonjourLocator,
            source: source
        )
    }
}

enum RemoteError: String, CaseIterable, Error, Sendable {
    case networkUnreachable = "NETWORK_UNREACHABLE"
    case tvNotFound = "TV_NOT_FOUND"
    case pairingRequired = "PAIRING_REQUIRED"
    case pairingCodeInvalid = "PAIRING_CODE_INVALID"
    case pairingRejected = "PAIRING_REJECTED"
    case pairingTimeout = "PAIRING_TIMEOUT"
    case trustChanged = "TRUST_CHANGED"
    case connectionLost = "CONNECTION_LOST"
    case voicePermissionDenied = "VOICE_PERMISSION_DENIED"
    case voiceSessionFailed = "VOICE_SESSION_FAILED"
    case textInputFailed = "TEXT_INPUT_FAILED"
    case unknown = "UNKNOWN"
}

enum RemoteState: Equatable, Sendable {
    case idle
    case discovering([TvCandidate])
    case connecting(RemoteDevice)
    case needsPairing(RemoteDevice)
    case pairing(RemoteDevice)
    case connected(RemoteDevice)
    case reconnecting(RemoteDevice, attempt: Int)
    case disconnected(RemoteDevice?)
    case failed(RemoteDevice?, reason: RemoteError, recoverable: Bool)
}

enum RemoteSessionEvent: Equatable, Sendable {
    case connected(RemoteDevice)
    case failed(RemoteDevice?, reason: RemoteError, recoverable: Bool)
}

enum ForegroundAction: Equatable {
    case none
    case connectRemembered
    case startDiscovery
}

enum ForegroundPolicy {
    static func action(alreadyActive: Bool, hasValidPairing: Bool) -> ForegroundAction {
        guard !alreadyActive else { return .none }
        return hasValidPairing ? .connectRemembered : .startDiscovery
    }

    static func allowsAutomaticDiscovery(isActive: Bool, hasValidPairing: Bool) -> Bool {
        isActive && !hasValidPairing
    }
}

enum BackgroundSessionAction: Equatable, Sendable {
    case disconnect
    case retainConnectedSession
}

enum BackgroundSessionPolicy {
    static func action(
        keepReadyEnabled: Bool,
        keepAliveAvailable: Bool,
        hasValidPairing: Bool,
        isConnected: Bool,
        disconnectedByUser: Bool
    ) -> BackgroundSessionAction {
        guard keepReadyEnabled,
              keepAliveAvailable,
              hasValidPairing,
              isConnected,
              !disconnectedByUser else {
            return .disconnect
        }
        return .retainConnectedSession
    }
}

enum KeepAliveStatus: Equatable, Sendable {
    case off
    case starting
    case ready
    case interrupted
}

enum ReconnectPolicy {
    static let delays: [TimeInterval] = [1, 2, 4]
}

enum PairingCodeValidator {
    static func isValid(_ code: String) -> Bool {
        normalized(code) != nil
    }

    static func normalized(_ code: String) -> String? {
        let value = code.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        guard value.utf8.count == 6,
              value.utf8.allSatisfy({ byte in
                  (48...57).contains(byte) || (65...70).contains(byte)
              }) else {
            return nil
        }
        return value
    }
}

@MainActor
protocol DiscoveryControlling: AnyObject {
    var onCandidatesChanged: (([TvCandidate]) -> Void)? { get set }
    func start()
    func stop()
}

@MainActor
protocol RemoteSessionControlling: AnyObject {
    var onEvent: ((RemoteSessionEvent) -> Void)? { get set }
    func connect(to record: LastTvRecord)
    func disconnect()
    func send(command: RemoteCommand, action: RemoteKeyAction)
}

@MainActor
protocol BackgroundKeepAliveControlling: AnyObject {
    var isAvailable: Bool { get }
    var status: KeepAliveStatus { get }
    var onStatusChanged: ((KeepAliveStatus) -> Void)? { get set }
    func start()
    func stop()
}

@MainActor
protocol ClientIdentityValidating: AnyObject {
    func status(matching fingerprint: String) throws -> ClientIdentityStatus
    func deleteIdentity() throws
}

enum ClientIdentityStatus: Equatable {
    case matches
    case missing
    case mismatch
}

extension IdentityStore: ClientIdentityValidating {}

protocol LastTvStoring: AnyObject {
    func load() throws -> LastTvRecord?
    func save(_ record: LastTvRecord) throws
    func clear()
}
